File: python_backend/game_logic/pathsInfo/get_path_data.py
from game_logic.pathsInfo.get_coord_path import _get_coord_path
from game_logic.pathsInfo.get_path import _get_path
from game_logic.pathsInfo.get_pieces_on_path import _get_pieces_on_path
from game_logic.pathsInfo.get_statuses import _get_statuses
from game_logic.pathsInfo.get_piece_matches_path import _get_piece_matches_path
from game_logic.step_funcs.step_funcs import *
from game_logic.printers.print_board import print_board
from game_logic.test_objects.get_standard_piece_defs import get_standard_piece_defs
from game_logic.test_objects.get_standard_id_dict import get_standard_id_dict
from game_logic.test_objects.sample_board_dicts import sample_board_dicts


def _get_path_data(board, sqr, color, piece_defs, id_dict, step_func):
    """get all info about a path on board determined by step function stemming from sqr"""
    x, y = sqr[0], sqr[1]
    coord_path = _get_coord_path(x, y, step_func)
    path = _get_path(board, x, y, step_func)
    piece_ids = _get_pieces_on_path(board, x, y, step_func)
    statuses = _get_statuses(piece_ids, color)
    piece_matches_path = []
    for piece_id in piece_ids:
        id_ = piece_id[1].lower()
        piece_name = id_dict[id_]
        piece_matches_path.append(_get_piece_matches_path(step_func, piece_defs, piece_name, color))
    return [coord_path, path, piece_ids, statuses, piece_matches_path]


if __name__ == "__main__":

    board = sample_board_dicts['check_example3']
    piece_defs = get_standard_piece_defs()
    id_dict = get_standard_id_dict()

    # test 1, 90:
    print("test 1:")
    coord_path, path, pieces, statuses, piece_matches_path = _get_path_data(board, (4, 4), 'W', piece_defs, id_dict, step_1sqr90d)
    print_board(board, heading="coord_path:{} path:{} pieces:{} statuses:{} piece_matches_path:{}".format(coord_path,
                                                                                                          path,
                                                                                                          pieces,
                                                                                                          statuses,
                                                                                                          piece_matches_path),
                                                                                                          highlights=coord_path,
                                                                                                          highlights3=[(4, 4)])
    print('')

    # test 2, 45:
    print('test 2:')
    coord_path, path, pieces, statuses, piece_matches_path = _get_path_data(board, (4, 4), 'W', piece_defs, id_dict, step_1sqr45d)
    print_board(board, heading="coord_path:{} path:{} pieces:{} statuses:{} piece_matches_path:{}".format(coord_path,
                                                                                                          path,
                                                                                                          pieces,
                                                                                                          statuses,
                                                                                                          piece_matches_path),
                                                                                                          highlights=coord_path,
                                                                                                          highlights3=[(4, 4)])
    print('')

    # test 3, 0:
    print('test 3:')
    coord_path, path, pieces, statuses, piece_matches_path = _get_path_data(board, (4, 4), 'W', piece_defs, id_dict, step_1sqr0d)
    print_board(board, heading="coord_path:{} path:{} pieces:{} statuses:{} piece_matches_path:{}".format(coord_path,
                                                                                                          path,
                                                                                                          pieces,
                                                                                                          statuses,
                                                                                                          piece_matches_path),
                                                                                                          highlights=coord_path,
                                                                                                          highlights3=[(4, 4)])
    print('')

    # test 4, 315:
    print('test 4:')
    coord_path, path, pieces, statuses, piece_matches_path = _get_path_data(board, (4, 4), 'W', piece_defs, id_dict, step_1sqr315d)
    print_board(board, heading="coord_path:{} path:{} pieces:{} statuses:{} piece_matches_path:{}".format(coord_path,
                                                                                                          path,
                                                                                                          pieces,
                                                                                                          statuses,
                                                                                                          piece_matches_path),
                                                                                                          highlights=coord_path,
                                                                                                          highlights3=[(4, 4)])
    print('')

    # test 5, 270:
    print('test 5:')
    coord_path, path, pieces, statuses, piece_matches_path = _get_path_data(board, (4, 4), 'W', piece_defs, id_dict, step_1sqr270d)
    print_board(board, heading="coord_path:{} path:{} pieces:{} statuses:{} piece_matches_path:{}".format(coord_path,
                                                                                                          path,
                                                                                                          pieces,
                                                                                                          statuses,
                                                                                                          piece_matches_path),
                                                                                                          highlights=coord_path,
                                                                                                          highlights3=[(4, 4)])
    print('')

    # test 6, 225:
    print('test 6:')
    coord_path, path, pieces, statuses, piece_matches_path = _get_path_data(board, (4, 4), 'W', piece_defs, id_dict, step_1sqr225d)
    print_board(board, heading="coord_path:{} path:{} pieces:{} statuses:{} piece_matches_path:{}".format(coord_path,
                                                                                                          path,
                                                                                                          pieces,
                                                                                                          statuses,
                                                                                                          piece_matches_path),
                                                                                                          highlights=coord_path,
                                                                                                          highlights3=[(4, 4)])
    print('')

    # test 7, 180:
    print('test 7:')
    coord_path, path, pieces, statuses, piece_matches_path = _get_path_data(board, (4, 4), 'W', piece_defs, id_dict, step_1sqr180d)
    print_board(board, heading="coord_path:{} path:{} pieces:{} statuses:{} piece_matches_path:{}".format(coord_path,
                                                                                                          path,
                                                                                                          pieces,
                                                                                                          statuses,
                                                                                                          piece_matches_path),
                                                                                                          highlights=coord_path,
                                                                                                          highlights3=[(4, 4)])
    print('')

    # test 8, 135:
    print('test 8:')
    coord_path, path, pieces, statuses, piece_matches_path = _get_path_data(board, (4, 4), 'W', piece_defs, id_dict, step_1sqr135d)
    print_board(board, heading="coord_path:{} path:{} pieces:{} statuses:{} piece_matches_path:{}".format(coord_path,
                                                                                                          path,
                                                                                                          pieces,
                                                                                                          statuses,
                                                                                                          piece_matches_path),
                                                                                                          highlights=coord_path,
                                                                                                          highlights3=[(4, 4)])
    print('')

File: app/src/game_logic/pathsInfo/getPathData.js
import {getCoordPath} from "../pathsInfo/getCoordPath";
import {getPath} from "../pathsInfo/getPath";
import {getPiecesOnPath} from "./getPiecesOnPath";
import {getStatuses} from "./getStatuses";
import {getPieceMatchesPath} from "./getPieceMatchesPath";
import {getStandardPieceDefs} from "../testObjects/getStandardPieceDefs";
import {getStandardIdDict} from "../testObjects/getStandardIdDict";
import {sampleBoardDicts} from "../testObjects/sampleBoardDicts";
import {mapListXyToRf} from "../coordType/mapListXyToRf";
import {printBoard} from "../printers/printBoard";
import {
    step_1sqr90d,
    step_1sqr45d,
    step_1sqr0d,
    step_1sqr315d,
    step_1sqr270d,
    step_1sqr225d,
    step_1sqr180d,
    step_1sqr135d,
} from "../helpers/stepFuncs";


export function getPathData(board, sqr, color, pieceDefs, idDict, stepFunc) {
    /*get all info about a path on board determined by step function stemming from sqr**/
    let [x, y] = [...sqr];
    var coordPath = getCoordPath(x, y, stepFunc);
    var path = getPath(board, x, y, stepFunc);
    var pieceIds = getPiecesOnPath(board, x, y, stepFunc);
    var statuses = getStatuses(pieceIds, color);
    var pieceMatchesPath = [];
    var id, pieceName;
    for (var pieceId of pieceIds) {
        id = pieceId[1].toLowerCase();
        pieceName = idDict[id];
        pieceMatchesPath.push(getPieceMatchesPath(stepFunc, pieceDefs, pieceName, color));
    }
    return [coordPath, path, pieceIds, statuses, pieceMatchesPath];
}

// module.exports = getPathData;

/** for node.js
if (require.main === module) {
    let board = sampleBoardDicts["check_example3"];
    let pieceDefs = getStandardPieceDefs();
    let idDict = getStandardIdDict();

    // test 1, 90:
    console.log("test 1:");
    var [coordPath, path, pieces, statuses, pieceMatchesPath] = getPathData(
        board,
        [4, 4],
        "W",
        pieceDefs,
        idDict,
        step_1sqr90d
    );
    printBoard(
        board,
        `path:${path} pieces:${pieces} statuses:${statuses} pieceMatchesPath:${pieceMatchesPath}`,
        mapListXyToRf(coordPath),
        ["d4"]
    );
    console.log("");


    // test 2, 45:
    console.log("test 2:")
    var [coordPath, path, pieces, statusVals, pieceMatchesPath] = getPathData(
        board,
        [4, 4],
        "W",
        pieceDefs,
        idDict,
        step_1sqr45d
    );
    printBoard(
        board,
        `path:${path} pieces:${pieces} statuses:${statusVals} pieceMatchesPath:${pieceMatchesPath}`,
        mapListXyToRf(coordPath),
        ["d4"]
    );
    console.log("");


    // test 3, 0:
    console.log("test 3:")
    var [coordPath, path, pieces, statuses, pieceMatchesPath] = getPathData(
        board,
        [4, 4],
        "W",
        pieceDefs,
        idDict,
        step_1sqr0d
    );
    printBoard(
        board,
        `path:${path} pieces:${pieces} statuses:${statuses} pieceMatchesPath:${pieceMatchesPath}`,
        mapListXyToRf(coordPath),
        ["d4"]
    );
    console.log("");

    
    // test 4, 315:
    // console.log("test 4:")
    var [coordPath, path, pieces, statuses, pieceMatchesPath] = getPathData(
        board,
        [4, 4],
        "W",
        pieceDefs,
        idDict,
        step_1sqr315d
    );
    printBoard(
        board,
        `path:${path} pieces:${pieces} statuses:${statuses} pieceMatchesPath:${pieceMatchesPath}`,
        mapListXyToRf(coordPath),
        ["d4"]
    );
    console.log("");


    // test 5, 270:
    console.log("test 5:")
    var [coordPath, path, pieces, statuses, pieceMatchesPath] = getPathData(
        board,
        [4, 4],
        "W",
        pieceDefs,
        idDict,
        step_1sqr270d
    );
    printBoard(
        board,
        `path:${path} pieces:${pieces} statuses:${statuses} pieceMatchesPath:${pieceMatchesPath}`,
        mapListXyToRf(coordPath),
        ["d4"]
    );
    console.log("");


    // test 6, 225:
    console.log("test 6:")
    var [coordPath, path, pieces, statuses, pieceMatchesPath] = getPathData(
        board,
        [4, 4],
        "W",
        pieceDefs,
        idDict,
        step_1sqr225d
    );
    printBoard(
        board,
        `path:${path} pieces:${pieces} statuses:${statuses} pieceMatchesPath:${pieceMatchesPath}`,
        mapListXyToRf(coordPath),
        ["d4"]
    );
    console.log("");


    // test 7, 180:
    console.log("test 7:")
    var [coordPath, path, pieces, statuses, pieceMatchesPath] = getPathData(
        board,
        [4, 4],
        "W",
        pieceDefs,
        idDict,
        step_1sqr180d
    );
    printBoard(
        board,
        `path:${path} pieces:${pieces} statuses:${statuses} pieceMatchesPath:${pieceMatchesPath}`,
        mapListXyToRf(coordPath),
        ["d4"]
    );
    console.log("");


    // test 8, 135:
    console.log("test 8:")
    var [coordPath, path, pieces, statuses, pieceMatchesPath] = getPathData(
        board,
        [4, 4],
        "W",
        pieceDefs,
        idDict,
        step_1sqr135d
    );
    printBoard(
        board,
        `path:${path} pieces:${pieces} statuses:${statuses} pieceMatchesPath:${pieceMatchesPath}`,
        mapListXyToRf(coordPath),
        ["d4"]
    );
    console.log("");

}
*/
